### PR TITLE
search: preserve scroll position on navigating back to search results 🎉

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

The scroll position is not preserved when going back to search results.

## Changes

**TLDR** The scroll position is now preserved for page refresh and back navigation 🎉 

This PR does two things:
- Adds `ScrollManager` that tracks the scroll position of the search results container, persists it to session storage, and restores it on back navigation or page refresh.
    - [react-scroll-manager](https://github.com/trevorr/react-scroll-manager) is used to handle scroll position info.
- To avoid multiple UI jumps required to reach the previous scroll position, we need to render all items that the user already scrolled through. It's done by extending the `itemsToShow` state used in `StreamingSearchResultsList`.
    - It tracks the number of items shown for the last query and persists this info to session storage. 
    - See `use-items-to-show.test.ts` with tests for all important use-cases.

https://user-images.githubusercontent.com/3846380/146541677-b48456d9-eaa5-42bd-a550-6fbeb97fc248.mov

Should be merged together with https://github.com/sourcegraph/sourcegraph/pull/29176.
Closes https://github.com/sourcegraph/sourcegraph/issues/28845.

### CODE REVIEW

**Review by olafurpg — APPROVED** [review]
I just tried this locally and it such a nice improvement 😍 Thank you @valerybugakov!

LGTM, but I didn't review the code in detail

File: client/web/src/search/results/StreamingSearchResultsList.tsx [hunk 1]
```diff
@@ -156,6 +139,13 @@ export const StreamingSearchResultsList: React.FunctionComponent<StreamingSearch
     )
 }
 
+function itemKey(item: SearchMatch): string {
```
Comment (vovakulikov) [inline client/web/src/search/results/StreamingSearchResultsList.tsx:142]: Nice. I'm wondering do we somehow enforce moving function out of render method in eslint

**Review by vovakulikov — APPROVED** [review]
Thanks, @valerybugakov this is a cool change!

File: client/web/src/search/results/StreamingSearchResultsList.tsx (hunk 1 above)
Comment (valerybugakov) [reply client/web/src/search/results/StreamingSearchResultsList.tsx:142]: It seems that the answer is no now, and it would be a good addition. 👍

**Review by limitedmage — APPROVED** [review]
This is awesome! Thanks for building this. I wonder if this would also work to preserve file scroll position globally, as mentioned in #24758?

**Review by novoselrok — APPROVED** [review]

**Comment by valerybugakov** [comment]
Yep, @limitedmage, we can use the same approach to address https://github.com/sourcegraph/sourcegraph/issues/24758.
We need to find time to look into it because the implementation will vary depending on the scrolling container on the page.
